File: rust/core/tests/drift_test.rs
//! Port of `DriftTest.java` / `drift.test.ts` (unit-gated; drift has no golden).

use std::collections::BTreeMap;
use varar_core::drift::{
    BaselineExample, BaselineStore, Drifted, OathBaseline, VarLock, derive_oath_baseline,
    detect_drift, live_examples, message, parse_var_lock, reconcile_drift, stringify_var_lock,
};
use varar_core::handler::Handler;
use varar_core::hash::hash_source;
use varar_core::parse::parse;
use varar_core::plan::{ExecutionPlan, plan};
use varar_core::registry::{Registry, add_step, create_registry};
use varar_core::span::Span;
use varar_core::step_kind::StepKind;

fn reg(with_step: bool) -> Registry {
    let r = create_registry();
    if with_step {
        add_step(&r, "I withdraw {int}", "steps.ts", 1, Handler::noop(), Some(StepKind::Stimulus))
            .unwrap()
    } else {
        r
    }
}

fn roman_reg(with_step: bool) -> Registry {
    let r = create_registry();
    if with_step {
        add_step(
            &r,
            "a decimal and a roman number",
            "steps.ts",
            1,
            Handler::noop(),
            Some(StepKind::Sensor),
        )
        .unwrap()
    } else {
        r
    }
}

fn plan_of(source: &str, r: &Registry) -> ExecutionPlan {
    plan(&parse("w.md", source), r)
}

fn bare(drifts: &[Drifted]) -> Vec<String> {
    drifts
        .iter()
        .map(|d| format!("{}@{}", d.name, d.line))
        .collect()
}

#[derive(Default)]
struct MemoryStore {
    contents: Option<String>,
}

impl BaselineStore for MemoryStore {
    fn read(&self) -> Option<String> {
        self.contents.clone()
    }
    fn write(&mut self, c: &str) {
        self.contents = Some(c.to_string());
    }
}

fn library_lock() -> VarLock {
    let mut oaths = BTreeMap::new();
    oaths.insert(
        "library.md".to_string(),
        OathBaseline {
            source_hash: "fnv1a:1a2b3c4d".to_string(),
            examples: vec![BaselineExample {
                name: "I check out".to_string(),
                line: 7,
            }],
        },
    );
    VarLock { version: 2, oaths }
}

#[test]
fn hash_matches_the_typescript_vectors() {
    assert_eq!("fnv1a:4f9f2cab", hash_source("hello"));
    assert_eq!("fnv1a:1a47e90b", hash_source("abc"));
    assert_eq!("fnv1a:4eace75e", hash_source("# Title\n"));
}

#[test]
fn live_examples_records_one_entry_per_example_producing_paragraph() {
    let var_doc = parse("w.md", "I withdraw 40.");
    assert_eq!(
        vec![BaselineExample {
            name: "I withdraw 40".to_string(),
            line: 1
        }],
        live_examples(&var_doc, &plan_of("I withdraw 40.", &reg(true)))
    );
}

#[test]
fn derive_oath_baseline_carries_the_fingerprint() {
    let source = "I withdraw 40.";
    let var_doc = parse("w.md", source);
    let baseline = derive_oath_baseline(source, &var_doc, &plan_of(source, &reg(true)));
    assert_eq!(hash_source(source), baseline.source_hash);
    assert_eq!(
        vec![BaselineExample {
            name: "I withdraw 40".to_string(),
            line: 1
        }],
        baseline.examples
    );
}

#[test]
fn no_baseline_means_no_drift() {
    let var_doc = parse("w.md", "I withdraw 40.");
    assert!(detect_drift(None, &var_doc, &plan_of("I withdraw 40.", &reg(true))).is_empty());
}

#[test]
fn a_renamed_step_drifts() {
    let source = "I withdraw 40.";
    let var_doc = parse("w.md", source);
    let baseline = derive_oath_baseline(source, &var_doc, &plan_of(source, &reg(true)));
    assert_eq!(
        vec!["I withdraw 40@1".to_string()],
        bare(&detect_drift(Some(&baseline), &var_doc, &plan_of(source, &reg(false))))
    );
}

#[test]
fn an_in_place_typo_drifts() {
    let before = "I withdraw 40.";
    let baseline =
        derive_oath_baseline(before, &parse("w.md", before), &plan_of(before, &reg(true)));
    let after = "I withdrraw 40.";
    let after_doc = parse("w.md", after);
    assert_eq!(
        vec!["I withdraw 40@1".to_string()],
        bare(&detect_drift(Some(&baseline), &after_doc, &plan_of(after, &reg(true))))
    );
}

#[test]
fn a_deleted_paragraph_is_not_drift() {
    let before = "I withdraw 40.";
    let baseline =
        derive_oath_baseline(before, &parse("w.md", before), &plan_of(before, &reg(true)));
    let after_doc = parse("w.md", "");
    assert!(detect_drift(Some(&baseline), &after_doc, &plan_of("", &reg(true))).is_empty());
}

#[test]
fn moving_and_rewording_a_still_matching_example_does_not_drift() {
    let before = "I withdraw 40.\n\nI withdraw 10.";
    let baseline =
        derive_oath_baseline(before, &parse("w.md", before), &plan_of(before, &reg(true)));
    let after = "I withdraw 11.\n\nI withdraw 40.";
    assert!(
        detect_drift(Some(&baseline), &parse("w.md", after), &plan_of(after, &reg(true)))
            .is_empty()
    );
}

#[test]
fn move_reword_prose_on_old_line_does_not_false_positive() {
    let before = "I withdraw 40.";
    let baseline =
        derive_oath_baseline(before, &parse("w.md", before), &plan_of(before, &reg(true)));
    let after = "Just some notes.\n\nI withdraw 41.";
    assert!(
        detect_drift(Some(&baseline), &parse("w.md", after), &plan_of(after, &reg(true)))
            .is_empty()
    );
}

#[test]
fn a_paragraph_rewritten_past_recognition_is_not_drift() {
    let before = "I withdraw 40.";
    let baseline =
        derive_oath_baseline(before, &parse("w.md", before), &plan_of(before, &reg(true)));
    let after = "The branch closed years ago.";
    assert!(
        detect_drift(Some(&baseline), &parse("w.md", after), &plan_of(after, &reg(true)))
            .is_empty()
    );
}

const ROMAN: &str = "Each row gives a decimal and a roman number:\n\n| decimal | roman |\n| ------: | :---- |\n| 3 | III |\n| 9 | IX |\n";

#[test]
fn header_bound_table_records_its_binding_paragraph_once() {
    let var_doc = parse("r.md", ROMAN);
    assert_eq!(
        vec![BaselineExample {
            name: "Each row gives a decimal and a roman number:".to_string(),
            line: 1
        }],
        live_examples(&var_doc, &plan(&var_doc, &roman_reg(true)))
    );
}

#[test]
fn a_header_bound_binding_paragraph_that_stops_matching_drifts() {
    let var_doc = parse("r.md", ROMAN);
    let baseline = derive_oath_baseline(ROMAN, &var_doc, &plan(&var_doc, &roman_reg(true)));
    assert_eq!(
        vec!["Each row gives a decimal and a roman number:@1".to_string()],
        bare(&detect_drift(Some(&baseline), &var_doc, &plan(&var_doc, &roman_reg(false))))
    );
}

#[test]
fn reconcile_records_then_reports_and_preserves_on_drift() {
    let source = "I withdraw 40.";
    let var_doc = parse("w.md", source);
    let mut store = MemoryStore::default();
    assert!(
        reconcile_drift(&mut store, "w.md", source, &var_doc, &plan_of(source, &reg(true)), false)
            .is_empty()
    );
    let before_lock = store.contents.clone();
    let drifts =
        reconcile_drift(&mut store, "w.md", source, &var_doc, &plan_of(source, &reg(false)), false);
    assert_eq!(vec!["I withdraw 40@1".to_string()], bare(&drifts));
    assert_eq!(before_lock, store.contents); // preserved while unacknowledged
}

#[test]
fn reconcile_update_mode_accepts_drift() {
    let source = "I withdraw 40.";
    let var_doc = parse("w.md", source);
    let mut store = MemoryStore::default();
    reconcile_drift(&mut store, "w.md", source, &var_doc, &plan_of(source, &reg(true)), false);
    assert!(
        reconcile_drift(&mut store, "w.md", source, &var_doc, &plan_of(source, &reg(false)), true)
            .is_empty()
    );
    let lock = parse_var_lock(store.contents.as_ref().unwrap()).unwrap();
    assert_eq!(Vec::<BaselineExample>::new(), lock.oaths.get("w.md").unwrap().examples);
}

const EXPECTED_LOCK: &str = "{\n  \"version\": 2,\n  \"oaths\": {\n    \"library.md\": {\n      \"sourceHash\": \"fnv1a:1a2b3c4d\",\n      \"examples\": [\n        {\n          \"name\": \"I check out\",\n          \"line\": 7\n        }\n      ]\n    }\n  }\n}\n";

#[test]
fn stringify_matches_the_typescript_serializer_byte_for_byte() {
    assert_eq!(EXPECTED_LOCK, stringify_var_lock(&library_lock()));
}

#[test]
fn parse_round_trips_a_valid_lock() {
    let parsed = parse_var_lock(&stringify_var_lock(&library_lock())).unwrap();
    assert_eq!("fnv1a:1a2b3c4d", parsed.oaths.get("library.md").unwrap().source_hash);
    assert_eq!(
        vec![BaselineExample {
            name: "I check out".to_string(),
            line: 7
        }],
        parsed.oaths.get("library.md").unwrap().examples
    );
}

#[test]
fn parse_rejects_malformed_input() {
    assert!(parse_var_lock("not json").is_none());
    assert!(parse_var_lock("{}").is_none());
    assert!(parse_var_lock("{\"version\":1,\"oaths\":{}}").is_none());
    assert!(parse_var_lock("{\"version\":2,\"oaths\":{\"a.md\":{\"examples\":[]}}}").is_none());
}

// ---- Merged examples keep per-paragraph drift granularity (ADR 0012) -------

fn deposit_withdraw_reg(with_deposit: bool) -> Registry {
    let mut r = create_registry();
    if with_deposit {
        r = add_step(
            &r,
            "I deposit {int}",
            "steps.ts",
            1,
            Handler::noop(),
            Some(StepKind::Stimulus),
        )
        .unwrap();
    }
    add_step(&r, "I withdraw {int}", "steps.ts", 2, Handler::noop(), Some(StepKind::Stimulus))
        .unwrap()
}

#[test]
fn two_paragraphs_that_merge_into_one_example_are_each_a_live_baseline_entry() {
    let source = "I deposit 100.\n\nI withdraw 40.";
    let var_doc = parse("w.md", source);
    let plan1 = plan(&var_doc, &deposit_withdraw_reg(true));
    // One planned example (the two paragraphs merged), but two live entries.
    assert_eq!(1, plan1.examples.len());
    assert_eq!(
        vec![
            BaselineExample {
                name: "I deposit 100".to_string(),
                line: 1
            },
            BaselineExample {
                name: "I withdraw 40".to_string(),
                line: 3
            },
        ],
        live_examples(&var_doc, &plan1)
    );
}

#[test]
fn deleting_one_step_def_of_a_merged_example_drifts_only_the_now_prose_paragraph() {
    let source = "I deposit 100.\n\nI withdraw 40.";
    let var_doc = parse("w.md", source);
    let baseline =
        derive_oath_baseline(source, &var_doc, &plan(&var_doc, &deposit_withdraw_reg(true)));
    // The deposit step is gone: its paragraph becomes prose, splitting the
    // example. The withdraw paragraph is still live; the deposit one drifts.
    let drift =
        detect_drift(Some(&baseline), &var_doc, &plan(&var_doc, &deposit_withdraw_reg(false)));
    assert_eq!(vec!["I deposit 100@1".to_string()], bare(&drift));
}

#[test]
fn drift_message_names_the_paragraph() {
    let d = Drifted {
        name: "I withdraw 40".to_string(),
        line: 1,
        span: Span::from_offsets("I withdraw 40.", 0, 13),
    };
    assert!(message(&d).contains("I withdraw 40"));
    assert!(!message(&d).trim().is_empty());
}

File: java/core/src/test/java/dev/varar/core/DriftTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** Translated from {@code var-core/tests/drift.test.ts} + hash.test.ts vectors. */
class DriftTest {

    private static final Object NOOP_HANDLER = (Runnable) () -> {};

    private static Registry reg(boolean withStep) {
        Registry r = Registry.createRegistry();
        if (withStep) {
            r = Registry.addStep(r, "I withdraw {int}", "steps.ts", 1, NOOP_HANDLER, StepKind.STIMULUS);
        }
        return r;
    }

    private static Registry romanReg(boolean withStep) {
        Registry r = Registry.createRegistry();
        if (withStep) {
            r = Registry.addStep(r, "a decimal and a roman number", "steps.ts", 1, NOOP_HANDLER, StepKind.SENSOR);
        }
        return r;
    }

    private static Plan.ExecutionPlan planOf(String source, Registry r) {
        return Plan.plan(Parse.parse("w.md", source), r);
    }

    private static List<String> bare(List<Drift.Drifted> drifts) {
        List<String> out = new ArrayList<>();
        for (Drift.Drifted d : drifts) out.add(d.name() + "@" + d.line());
        return out;
    }

    /** An in-memory BaselineStore. */
    private static final class MemoryStore implements Drift.BaselineStore {
        String contents;

        @Override
        public String read() {
            return contents;
        }

        @Override
        public void write(String c) {
            contents = c;
        }
    }

    @Test
    void hashMatchesTheTypescriptVectors() {
        assertEquals("fnv1a:4f9f2cab", Hash.hashSource("hello"));
        assertEquals("fnv1a:1a47e90b", Hash.hashSource("abc"));
        assertEquals("fnv1a:4eace75e", Hash.hashSource("# Title\n"));
    }

    @Test
    void liveExamplesRecordsOneEntryPerExampleProducingParagraph() {
        Ast.VarDoc varDoc = Parse.parse("w.md", "I withdraw 40.");
        assertEquals(
                List.of(new Drift.BaselineExample("I withdraw 40", 1)),
                Drift.liveExamples(varDoc, planOf("I withdraw 40.", reg(true))));
    }

    @Test
    void deriveOathBaselineCarriesTheFingerprint() {
        String source = "I withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        Drift.OathBaseline baseline = Drift.deriveOathBaseline(source, varDoc, planOf(source, reg(true)));
        assertEquals(Hash.hashSource(source), baseline.sourceHash());
        assertEquals(List.of(new Drift.BaselineExample("I withdraw 40", 1)), baseline.examples());
    }

    @Test
    void noBaselineMeansNoDrift() {
        Ast.VarDoc varDoc = Parse.parse("w.md", "I withdraw 40.");
        assertTrue(Drift.detectDrift(null, varDoc, planOf("I withdraw 40.", reg(true)))
                .isEmpty());
    }

    @Test
    void aRenamedStepDrifts() {
        String source = "I withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        Drift.OathBaseline baseline = Drift.deriveOathBaseline(source, varDoc, planOf(source, reg(true)));
        assertEquals(List.of("I withdraw 40@1"), bare(Drift.detectDrift(baseline, varDoc, planOf(source, reg(false)))));
    }

    @Test
    void anInPlaceTypoDrifts() {
        String before = "I withdraw 40.";
        Drift.OathBaseline baseline =
                Drift.deriveOathBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        String after = "I withdrraw 40.";
        Ast.VarDoc afterDoc = Parse.parse("w.md", after);
        assertEquals(List.of("I withdraw 40@1"), bare(Drift.detectDrift(baseline, afterDoc, planOf(after, reg(true)))));
    }

    @Test
    void aDeletedParagraphIsNotDrift() {
        String before = "I withdraw 40.";
        Drift.OathBaseline baseline =
                Drift.deriveOathBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        Ast.VarDoc afterDoc = Parse.parse("w.md", "");
        assertTrue(Drift.detectDrift(baseline, afterDoc, planOf("", reg(true))).isEmpty());
    }

    @Test
    void movingAndRewordingAStillMatchingExampleDoesNotDrift() {
        String before = "I withdraw 40.\n\nI withdraw 10.";
        Drift.OathBaseline baseline =
                Drift.deriveOathBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        String after = "I withdraw 11.\n\nI withdraw 40.";
        assertTrue(Drift.detectDrift(baseline, Parse.parse("w.md", after), planOf(after, reg(true)))
                .isEmpty());
    }

    @Test
    void moveRewordProseOnOldLineDoesNotFalsePositive() {
        String before = "I withdraw 40.";
        Drift.OathBaseline baseline =
                Drift.deriveOathBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        String after = "Just some notes.\n\nI withdraw 41.";
        assertTrue(Drift.detectDrift(baseline, Parse.parse("w.md", after), planOf(after, reg(true)))
                .isEmpty());
    }

    @Test
    void aParagraphRewrittenPastRecognitionIsNotDrift() {
        String before = "I withdraw 40.";
        Drift.OathBaseline baseline =
                Drift.deriveOathBaseline(before, Parse.parse("w.md", before), planOf(before, reg(true)));
        String after = "The branch closed years ago.";
        assertTrue(Drift.detectDrift(baseline, Parse.parse("w.md", after), planOf(after, reg(true)))
                .isEmpty());
    }

    private static final String ROMAN =
            "Each row gives a decimal and a roman number:\n\n| decimal | roman |\n| ------: | :---- |\n| 3 | III |\n| 9 | IX |\n";

    @Test
    void headerBoundTableRecordsItsBindingParagraphOnce() {
        Ast.VarDoc varDoc = Parse.parse("r.md", ROMAN);
        assertEquals(
                List.of(new Drift.BaselineExample("Each row gives a decimal and a roman number:", 1)),
                Drift.liveExamples(varDoc, Plan.plan(varDoc, romanReg(true))));
    }

    @Test
    void aHeaderBoundBindingParagraphThatStopsMatchingDrifts() {
        Ast.VarDoc varDoc = Parse.parse("r.md", ROMAN);
        Drift.OathBaseline baseline = Drift.deriveOathBaseline(ROMAN, varDoc, Plan.plan(varDoc, romanReg(true)));
        assertEquals(
                List.of("Each row gives a decimal and a roman number:@1"),
                bare(Drift.detectDrift(baseline, varDoc, Plan.plan(varDoc, romanReg(false)))));
    }

    @Test
    void reconcileRecordsThenReportsAndPreservesOnDrift() {
        String source = "I withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        MemoryStore store = new MemoryStore();
        assertTrue(Drift.reconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(true)), false)
                .isEmpty());
        String beforeLock = store.contents;
        List<Drift.Drifted> drifts =
                Drift.reconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(false)), false);
        assertEquals(List.of("I withdraw 40@1"), bare(drifts));
        assertEquals(beforeLock, store.contents); // preserved while unacknowledged
    }

    @Test
    void reconcileUpdateModeAcceptsDrift() {
        String source = "I withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        MemoryStore store = new MemoryStore();
        Drift.reconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(true)), false);
        assertTrue(Drift.reconcileDrift(store, "w.md", source, varDoc, planOf(source, reg(false)), true)
                .isEmpty());
        Drift.VarLock lock = Drift.parseVarLock(store.contents);
        assertEquals(List.of(), lock.oaths().get("w.md").examples());
    }

    private static final String EXPECTED_LOCK = "{\n"
            + "  \"version\": 2,\n"
            + "  \"oaths\": {\n"
            + "    \"library.md\": {\n"
            + "      \"sourceHash\": \"fnv1a:1a2b3c4d\",\n"
            + "      \"examples\": [\n"
            + "        {\n"
            + "          \"name\": \"I check out\",\n"
            + "          \"line\": 7\n"
            + "        }\n"
            + "      ]\n"
            + "    }\n"
            + "  }\n"
            + "}\n";

    @Test
    void stringifyMatchesTheTypescriptSerializerByteForByte() {
        Drift.VarLock lock = new Drift.VarLock(
                2,
                Map.of(
                        "library.md",
                        new Drift.OathBaseline(
                                "fnv1a:1a2b3c4d", List.of(new Drift.BaselineExample("I check out", 7)))));
        assertEquals(EXPECTED_LOCK, Drift.stringifyVarLock(lock));
    }

    @Test
    void parseRoundTripsAValidLock() {
        Drift.VarLock lock = new Drift.VarLock(
                2,
                Map.of(
                        "library.md",
                        new Drift.OathBaseline(
                                "fnv1a:1a2b3c4d", List.of(new Drift.BaselineExample("I check out", 7)))));
        Drift.VarLock parsed = Drift.parseVarLock(Drift.stringifyVarLock(lock));
        assertEquals("fnv1a:1a2b3c4d", parsed.oaths().get("library.md").sourceHash());
        assertEquals(
                List.of(new Drift.BaselineExample("I check out", 7)),
                parsed.oaths().get("library.md").examples());
    }

    @Test
    void parseRejectsMalformedInput() {
        assertNull(Drift.parseVarLock("not json"));
        assertNull(Drift.parseVarLock("{}"));
        assertNull(Drift.parseVarLock("{\"version\":1,\"oaths\":{}}"));
        assertNull(Drift.parseVarLock("{\"version\":2,\"oaths\":{\"a.md\":{\"examples\":[]}}}"));
    }

    @Test
    void driftMessageNamesTheParagraph() {
        Drift.Drifted d = new Drift.Drifted("I withdraw 40", 1, Span.spanFromOffsets("I withdraw 40.", 0, 13));
        assertTrue(Drift.message(d).contains("I withdraw 40"));
        assertFalse(Drift.message(d).isBlank());
    }

    // ---- Merged examples keep per-paragraph drift granularity (ADR 0012) -------

    private static Registry depositWithdrawReg(boolean withDeposit) {
        Registry r = Registry.createRegistry();
        if (withDeposit) {
            r = Registry.addStep(r, "I deposit {int}", "steps.ts", 1, NOOP_HANDLER, StepKind.STIMULUS);
        }
        r = Registry.addStep(r, "I withdraw {int}", "steps.ts", 2, NOOP_HANDLER, StepKind.STIMULUS);
        return r;
    }

    @Test
    void twoParagraphsThatMergeIntoOneExampleAreEachRecordedAsALiveBaselineEntry() {
        String source = "I deposit 100.\n\nI withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        Plan.ExecutionPlan plan = Plan.plan(varDoc, depositWithdrawReg(true));
        // One planned example (the two paragraphs merged), but two live entries.
        assertEquals(1, plan.examples().size());
        assertEquals(
                List.of(new Drift.BaselineExample("I deposit 100", 1), new Drift.BaselineExample("I withdraw 40", 3)),
                Drift.liveExamples(varDoc, plan));
    }

    @Test
    void deletingOneStepDefOfAMergedExampleDriftsOnlyTheNowProseParagraph() {
        String source = "I deposit 100.\n\nI withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        Drift.OathBaseline baseline =
                Drift.deriveOathBaseline(source, varDoc, Plan.plan(varDoc, depositWithdrawReg(true)));
        // The deposit step is gone: its paragraph becomes prose, splitting the example. The
        // withdraw paragraph is still live; the deposit one drifts.
        List<Drift.Drifted> drift = Drift.detectDrift(baseline, varDoc, Plan.plan(varDoc, depositWithdrawReg(false)));
        assertEquals(List.of("I deposit 100@1"), bare(drift));
    }

    // ---- Pruning baselines for oaths that no longer exist (issue #70) ----

    /**
     * A lock carrying two oaths, one of which the docs globs no longer match — the state a deleted
     * or moved .md leaves behind.
     */
    private static String lockWithStalePath() {
        String source = "I withdraw 40.";
        Ast.VarDoc varDoc = Parse.parse("w.md", source);
        Drift.OathBaseline baseline = Drift.deriveOathBaseline(source, varDoc, Plan.plan(varDoc, reg(true)));
        return Drift.stringifyVarLock(new Drift.VarLock(2, Map.of("varar/w.md", baseline, "w.md", baseline)));
    }

    @Test
    void pruneVarLockKeepsOnlyThePathsItIsGiven() {
        Drift.VarLock lock = Drift.parseVarLock(lockWithStalePath());
        Drift.VarLock pruned = Drift.pruneVarLock(lock, List.of("varar/w.md"));
        assertEquals(List.of("varar/w.md"), new ArrayList<>(pruned.oaths().keySet()));
    }

    @Test
    void pruneReportsStalePathsWithoutUpdateAndDoesNotWrite() {
        MemoryStore store = new MemoryStore();
        store.contents = lockWithStalePath();
        String before = store.contents;

        assertEquals(List.of("w.md"), Drift.pruneBaselines(store, List.of("varar/w.md"), false));
        // Reporting is not deleting: nothing is removed behind the author's back.
        assertEquals(before, store.contents);
    }

    @Test
    void pruneDropsStalePathsUnderUpdate() {
        MemoryStore store = new MemoryStore();
        store.contents = lockWithStalePath();

        assertEquals(List.of("w.md"), Drift.pruneBaselines(store, List.of("varar/w.md"), true));
        assertEquals(
                List.of("varar/w.md"),
                new ArrayList<>(Drift.parseVarLock(store.contents).oaths().keySet()));
    }

    @Test
    void pruneLeavesALockWithNoStalePathsUntouched() {
        MemoryStore store = new MemoryStore();
        store.contents = lockWithStalePath();
        String before = store.contents;

        assertTrue(
                Drift.pruneBaselines(store, List.of("varar/w.md", "w.md"), true).isEmpty());
        // Byte-identical, not merely equivalent — an unnecessary rewrite would show up as a
        // spurious diff in every consumer's working tree.
        assertEquals(before, store.contents);
    }

    @Test
    void pruneIsANoOpWhenThereIsNoBaselineYet() {
        MemoryStore store = new MemoryStore();
        assertTrue(Drift.pruneBaselines(store, List.of("varar/w.md"), true).isEmpty());
        assertNull(store.contents);
    }
}
